### ISSUE TITLE
On supporting non-github release('d) cores

### ISSUE THREAD

**Issue by neil-morrison44** [joshcampbell191/openfpga-cores-inventory#26]
As per the [Snow Bros 2 core](https://github.com/joshcampbell191/openfpga-cores-inventory/issues/25) & following on from the chat in the discord.

(I'll just talk about the output JSON in this rather than the input yml)

I _think_ what could work is something along the lines of (forgive the typescript notation):

```ts
type CoreRepo = {
  platform: "github",
  owner: string,
  repo: string
} | {
  platform: "github-within-tree",
  owner: string,
  repo: string,
  path: string, // "releases" for the snowbros 2 core
  filename: string // "pram0d.snowbros2_*" for the snowbros 2 core
} | {
  platform: "bitbucket", // just shoving this in here as another example
  owner: string,
  repo: string
}
```

Then in updaters they'll do their current behaviour for `platform: "github"`, and switch to a new download method for `"github-within-tree"` (which is a rubbish name, there'll be a better one...)

The "new method" (for whatever better name `github-within-tree` ends up getting) would compose the API call for https://api.github.com/repos/psomashekar/pram0d-pocket-dist-public/contents/releases then find the most recent file which matches the `filename` glob & download it (which isn't actually as different from using the releases API as I thought it would be...)

There's a couple of terminology things that might change since `release.tag_name` doesn't apply since it's not tagged & `repository` itself might make more sense as `download` but that's small fry stuff.

**Comment by goronfreeman** [comment]
Thanks for writing this up! I've been talking a bit more with @mattpannella, and our current position is that using GitHub releases is the standard that we want to stick with, at least for now. If someone releases a core on something like GitLab, I think it makes sense for us to support that, and GitLab has its own version of releases that should be compatible with how we're set up currently.

Unless more developers choose not to use releases for whatever reason, I'm not sure we want to add complexity and sink time into this for a single repo.

**Comment by neil-morrison44** [comment]
Seems a bit of a shame to be the inventory for _almost_ everything, but I can see how difficult it'd be given the format of the yml files etc

**Comment by goronfreeman** [comment]
I decided to look at this again and see if I could square it in my head, but I don't think I can get there still. One big question I have, is how do we know when there is a new release if there is no git tag? There's only a single `develop` branch in the repo. The only way we could handle this would be to re-download the ZIP every single time the script runs. We could _maybe_ look at the file creation date of the ZIP, but that's different across platforms.

We would also have to maintain two separate ways of parsing out information. We switched to using the GitHub API response to get the release date and core version info after initially going off of the JSON files in the cores, because we found that core authors don't do things in a standard way. For instance, the Space War core listed the version and release date of the original game, not the core.

I agree that it's kind of a shame we can't include these, but we're looking at maybe dozens of collective hours changing the API and updaters versus the core author just adding a release.

**Comment by neil-morrison44** [comment]
I _suspect_ what's likely to happen is the .zip files will be added to that repo with a new date appended (so you'll have multiples of each core with different dates appended). You'd _probably_ need to look at the date on the zip (or, yeah, download it and inspect inside) - or find the most recent & check the git hash.

The author might be amenable to adding the semver into the filename `snowbros2_20221019.1.0.0.zip` style, which would solve that without them having to do releases.

Yeah - I had the same issue with the SpaceWar core when I wrote the first JSON, pretty annoying that Analogue immediately broke their own rule in the example core...

**Comment by mattpannella** [comment]
re: checking the SHA, i think pram0d said he plans to add a new zip for each release and add a date to the file
pram0d.snowbros2_20221019.zip
pram0d.snowbros2_20221020.zip

etc

so i think what you'd need is to do is look for any file named `{core_name}*` and then parse the date strings to find the most recent one

**Comment by goronfreeman** [comment]
Bleh. Well, at least that's something to work with potentially. I got bored and very roughly started extracting out the current script as its own library. Working to see what it would look like to make this more modular, so we could plug in different classes to handle different things. I might keep experimenting with this as I get time/motivation, with the end goal being supporting GitLab as well, and if it's not too terrible, maybe the pram0d stuff.

**Comment by goronfreeman** [comment]
Another update on this. @mattpannella pointed out that Jotego releases his Mister cores in a mono repo as well. It is also the same structure as pram0d. With his cores being publicly released in the near future, I think it makes sense to put in the effort to support them.

**Comment by mattpannella** [comment]
was thinking about this a little bit, going off @neil-morrison44's structure in the op

```yml
- username: agg23
  cores:
    - display_name: Atari's 1972 Pong
      repository: openfpga-pong
- username: psomashekar
  cores:
    - display_name: Snow Bros 2
      repository: pram0d-pocket-dist-public
      mono-repo-directory: releases
      filename: pram0d.snowbros2
```
so if the mono-repo-directory is set (can pick a prettier name) the cronjob knows to look for releases in that directory of the repo, with the filename as the pattern to check for. and each release will have _DATE tacked on that we can use for versioning. (could also be semver as well. i guess as long as we assume everything after the `_` is the version, the comparison can be handled by the updater. because the cronjob can just grab the most recent file that matches the filename
and the filename can(should?) be the core name

and then json output could look something like. would only have a release, since there's no clean way to have pre-release.:

```json
  {
    "identifier": "agg23.Pong",
    "platform": "Pong",
    "repository": {
      "platform": "github",
      "owner": "agg23",
      "name": "openfpga-pong"
    },
    "release": {
      "tag_name": "1.2.0",
      "release_date": "2022-09-09T01:56:30Z",
      "platform": {
        "category": "Arcade",
        "name": "Pong",
        "manufacturer": "Atari",
        "year": 1972
      },
      "assets": []
    }
  },
  {
    "identifier": "pram0d.snowbros2",
    "platform": "Snow Bros 2",
    "repository": {
      "platform": "github",
      "owner": "psomashekar",
      "name": "pram0d-pocket-dist-public",
      "mono-repo-directory": "releases",
      "filename": "pram0d.snowbros2"
    },
    "release": {
      "version": "20221019",
      "release_date": "2022-10-07T18:27:05Z",
      "platform": {
        "category": "Arcade",
        "name": "Snow Bros 2",
        "manufacturer": "i dont know",
        "year": 1983
      },
      "assets": [
        {
           "platform": "snowbros2", 
           "filename":"snowbros2.rom"
        }
      ]
    }
  }
```

**Comment by neil-morrison44** [comment]
I would advise having the `repository.platform` property be the sole thing that the updaters use to key off how they should download the core - this one's of a similar (or perhaps more) difference from the "normal" `"github"` one as a bitbucket or gitlab release would be & it'd allow the updaters to just check `repository.platform` to know if they can handle it or not -- also neater to add to later if others release cores in even more unexpected ways.

-- but that'd mean big changes to the .yml files (but they might end up in a better place after the changes)

**Comment by joshcampbell191** [comment]
I'll take some time this week to hash things out and come up with a generic solution that can be applied the repositories va releases.

**Comment by joshcampbell191** [comment]
@neil-morrison44 @mattpannella @goronfreeman 

Sorry for the delay, I've been a little caught up in things. I also wanted to take the time to go through and not jump to any conclusions while trying to keep things flexible.

I'd like to propose adding a new section to the [repos.yml](https://github.com/joshcampbell191/openfpga-cores-inventory/blob/master/_data/repos.yml) under each core:
```yml
distribution:
  type: ""            # Mandatory - Can be one of the following: release, contents or url
  release:            # Optional - Contains the configurations for type 'release'
    asset_prefix: ""  # Optional - If specified, the prefix will be used to filter release assets. If multiple files match, only the first will be returned. If omitted, the first result will be returned.
  contents:           # Optional - Contains the configurations for type 'contents'
    path: ""          # Optional - If specified, the path will be used to resolve the directory / file. If omitted, the content of the repository's root will be used instead.
    file_prefix: ""   # Optional - If specified, the prefix will be used to filter repository contents. If multiple files match, only the first will be returned. If omitted, the first result will be returned.
  url:                # Optional - Contains the configurations for type 'url'
    release_date: ""  # Mandatory - The release date of the core.
    link: ""          # Mandatory - The URL to download the core.
```

Here's a brief explanation of the different distribution types:
- `release`: This is the method currently used by most of the cores hosted on GitHub.
- `contents`: This method is currently used by the [pram0d cores](https://github.com/psomashekar/pram0d-pocket-dist-public). 
- `url`: This method would allow cores to be hosted elsewhere / allow us to quickly support new cores while we iron out the details.

Some of the changes listed above will require some modifications to the logic used to retrieve the `release_date`.

Furthermore, I'd also like to add a `url` property to the release / prerelease property in the [cores.json](https://github.com/joshcampbell191/openfpga-cores-inventory/blob/master/api/v1/analogue-pocket/cores.json). This would allow updaters to completely bypass GitHub API (for the `release` and `contents` distribution types) and download cores directly.

```json
"release": {
  "tag_name": "0.9.0",
  "release_date": "2022-09-03T20:45:09Z",
  "url": "https://github.com/agg23/openfpga-arduboy/releases/download/0.9.0/agg23.Arduboy.0.9.0.zip",
  "platform": {
    "category": "Handheld",
    "name": "Arduboy",
    "manufacturer": "Arduboy",
    "year": 2016,
   },
   "assets": [
     {
      "platform": "arduboy",
      "extensions": [
        "hex"
      ]
    }
  ]
}
```

I believe these changes should be sufficient to cover all the scenarios that we've currently encountered, however, I'm looking to you guys in case I missed anything.

**Comment by mattpannella** [comment]
this all looks really great to me
i think the only thing that jumps out at me that could bring up an issue is including the url to the github release. 
1. this isn't terrible, cuz i'm already handling in my updater, but you have to iterate through the assets if there's multiple and make an educated guess on which zip is the actual core. i just check for a folder named "Cores" in the zip and assume thats safe enough to guess that its the actual release zip i want
2. the more annoying one: multiple cores in one repo. aka the gb/gbc repo haha. there's not a single release zip there

**Comment by joshcampbell191** [comment]
The `url` property would point directly to the core's archive. For repositories that host multiple cores, the `*_prefix` property can be used to filter the correct file.

**Comment by mattpannella** [comment]
oh duh, ok yep. thats great i like it. not sure if there's anything else we'd really need?

**Comment by goronfreeman** [comment]
I need to give this one more good read through when I'm a little less frazzled, but I think this makes sense. I've thought before that we should just add the URL directly, so I think that's a great addition here. I have a rough plan in place to make this happen. First, I think this is getting complex enough that instead of cramming all the code into the `workflows` directory in this project, it should be extracted out.

I created the new repo [corekeeper](https://github.com/goronfreeman/corekeeper) where I'm going to completely rewrite the logic for this and package it as a rubygem that we can easily include in this repo. My hope is that I can make the mode more modular, so we can plugin classes here and there to give us special behavior. This should be especially nice once we have to support another git hosting site. We can transfer the gem to the organization whenever we get around to creating that.

I want to get this gem spitting out an identical JSON file to what we're currently producing, then once that is working, move on to adding the new functionality that @joshcampbell191 described above.

Thanks for once again putting in the legwork to come up with a clean structure for us to start building on! 😃

**Comment by joshcampbell191** [comment]
@neil-morrison44 / @mattpannella does this approach work for you guys?

**Comment by neil-morrison44** [comment]
Think so yeah, so the existing cores would get something like
```json
distribution: {
  type: "url"
  url: {
    release_date: "2022-09-03T20:45:09Z",
    link: "https://github.com/agg23/openfpga-arduboy/releases/download/0.9.0/agg23.Arduboy.0.9.0.zip"
  }
}
```

and the snowbros core (etc) would get something like

```json
distribution: {
  type: "contents",
  contents: {
    path: "releases/"
    prefix: "pram0d.snowbros2_"
  }
```
If I'm understanding correctly?

**Comment by neil-morrison44** [comment]
Ah wait - confusion between yml & json, I see what you're saying now, they'd all get output as the `url` on the JSON side.

The point about the release notes still stands I think, I'd still like to know something about where the core lives (at least if it's on github with the release system)

**Comment by mattpannella** [comment]
the repo info would still be in the api response i think. so that wouldnt change. this is just additional, right?

**Comment by mattpannella** [comment]
lol should we discuss changing this to support jotegos stuff or just make that a new issue

**Comment by neil-morrison44** [comment]
I reckon there’ll probably be an additional step (mirror repo / building the zips somewhere / some sort of transformation) in between JOTEGO’s cores and this so I think a new issue probably makes sense & it’ll just feed into the `url` system discussed here

**Comment by neil-morrison44** [comment]
The JOTGEO cores should, hopefully, be going up in JTBIN soon as individual zips with their own version numbers (and no date suffix) - I think what we discussed for the pram0d cores _should_ work

**Comment by joshcampbell191** [comment]
@neil-morrison44 @mattpannella 

I know it's been a little while, but I'm just about finished with this feature.

I noticed that if we go with this approach, the `tag_name` field will be empty for cores that come from a `url` or a repository's `contents`. After some thinking, I was wondering if we can get rid of this in favor of the `core.metadata.version` field from `core.json`. However, some cores (JOTEGO) don't seem to implement this.

What are your thoughts?

**Comment by mattpannella** [comment]
Neil setup a GitHub action to inject the version into jotegos cores and save them as zips
https://github.com/jotego/jtbin/tree/master/pocket/zips

and yeah using the version is ideal. I think we already switched to that instead of tag in the current API version

**Comment by joshcampbell191** [comment]
Sounds like we need to revisit the `cores.json` format another time. I'll see if I can put something together shortly and ensure it meets everyone's needs.

**Comment by neil-morrison44** [comment]
Yeah that workflow should be putting a version into the JOTEGO cores now (it's just the git hash of the commit that last updated something within the core).

There is a problem with the workflow though where it replaces the zips even without any changes - I assumed it'd create the same zip file & git would ignore it but that's not the case (not sure if it's not the same zip file or if it's a timestamp / metadata thing)

That might cause issues, not sure.

& yeah, I don't think Pocket Sync looks at the tag name anymore, just the version so that's fine

**Comment by mattpannella** [comment]
@neil-morrison44 i think the issue in the jt release zips is it's adding the latest commit hash to every single core, so thats why its creating a new zip for every one
i just ran the v2 api in my updater it every jt core updated with the same version 

> Update Process Complete
Cores Updated:
jotego.jtcontra 64d5ae0
jotego.jtdd 64d5ae0
jotego.jtdd2 64d5ae0
jotego.jtgng 64d5ae0
jotego.jtkarnov 64d5ae0
jotego.jtkicker 64d5ae0
jotego.jtkunio 64d5ae0
jotego.jtmikie 64d5ae0
jotego.jtpang 64d5ae0
jotego.jtpinpon 64d5ae0
jotego.jtroadf 64d5ae0
jotego.jtroc 64d5ae0
jotego.jtsbaskt 64d5ae0
jotego.jttrack 64d5ae0
jotego.jtvigil 64d5ae0
jotego.jtyiear 64d5ae0

**Comment by neil-morrison44** [comment]
Oh - it _should be_ looking at the latest commit hash that affects the files which end up in the zip rather than just the latest commit hash full stop.

Could be something to do with the depth of the checkout, if that makes the logs useless.

(Also, v2 is up?)

**Comment by mattpannella** [comment]
josh said he's still got some stuff to tweak, but you can start testing
https://github.com/openfpga-cores-inventory/analogue-pocket/commit/baeb51c8201201996edac17edb5783f3d8148fc0

**Comment by joshcampbell191** [comment]
v2 of the API is now up and ready for testing. You can find the documentation here: https://github.com/openfpga-cores-inventory/analogue-pocket/blob/main/_docs/v2.md

Note: v0 and v1 of the API will only support cores that are distributed via the GitHub releases. 

Let me know if you come across any issues or would like to make changes.

**Comment by neil-morrison44** [comment]
@joshcampbell191 Trying it out in Pocket Sync now & it works perfectly, absolutely no issues - can see & install all the JOTEGO, pram0d cores etc.

Lots of my nasty code, including some ugly regex stuff for finding the download zip goes in the bin so that's great too.

Assuming @mattpannella finds the same are you reading for our updaters to start using V2?

**Comment by mattpannella** [comment]
yup i'm good to go on this end

**Comment by joshcampbell191** [comment]
I was just waiting on you both to confirm things are working on your end before locking in the changes.

**Comment by joshcampbell191** [comment]
@neil-morrison44 now that you mention it, I don't have a good solution for the pram0d cores (specifically garegga) since the logic I wrote for the file prefix captures the first match. I don't do any ordering by published dates since this could return incorrect results depending on how assets are uploaded / modified.

**Comment by neil-morrison44** [comment]
Is that where garegga & kingdmgp & sstriker are all included in one zip?
I noticed that, it isn't _too_ much of an issue for Pocket Sync but I guess it'd be hard to know which core the `version` number etc relates to + it's a little annoying that the user has to manually deselect the other files in the zip preview.

Not sure how to solve that beyond splitting the zips per core & re-hosting them somewhere but that's a bunch of work & ongoing file hosting.

(The Pocket Sync version that talks to V2 is going up just now by the way)

**Comment by joshcampbell191** [comment]
Yes, that's the one. 

The fact that they are packaged together doesn't matter on my end. My only issue with it is that there's multiple versions of the zip in the release folder and my prefix approach returns the first match based on the name. I chose that instead of GitHub dates since dates can be unreliable depending on when the assets are updated.

**Comment by neil-morrison44** [comment]
Got you, yeah no better solution to that comes to mind - first by filename should work assuming it's timestamps & numerically descending.
Looks like there's a `pram0d.garegga_20221029.zip` & a `pram0d.garegga_20221030.zip` just now & the API shows the `20221030.zip` one so hopefully that holds up long term.

---

On the zips with multiple cores thing - which one gets chosen as the one you take the `version` string etc from?

**Comment by mattpannella** [comment]
what if we further complicate the repo yaml config and add a flag for when there will be multiple zip releases (with dates) per core.
 then if that's true you grab all files that match your prefix filter instead of just one and sort them to get the latest based on the date.
it's pretty gross but it would work haha

**Comment by joshcampbell191** [comment]
Let's close this for now and leave things as is. Either way, the issue would be on my end and shouldn't stop you.
